docs/rfc/rfc-1.adoc

```diff
@@ -0,0 +1,102 @@
+understanding how Hierarchical Deterministic Wallets work. The basic concept is
+that you have a root private key and public key that allows you to supply an
+index that creates another private key + public key. To experiment with this,
+check out https://iancoleman.io/bip39/[bip39]. Conventionally, a derivation path is written as
+`m / purpose' / coin_type' / account' / change / address_index`. A `'` denotes
+that the index is "hardened", which means that a base of `2147483648` has been
+added to the address (and that the private key is required to derive the
+address). The HD wallet standard is blockchain agnostic, and different
+standards have been adopted for each community. Bitcoin, for example, always
+uses a purpose of `44'`, a coin type of `0`, and a change of `0` (the change
+address denotes where leftover bitcoin goes after being sent from the address).
+This means that for our purposes, we can vary the `account` and `address_index`
+fields: `m/44'/0/x/0/y`. Each path has `2^31` available hardened addresses,
+which means there are `2^62` unique hardened paths.
+
+The main attack that we want to avoid is someone monitoring children of the
+active wallet for deposits, and then telling the ethereum chain that they
+deposited the bitcoin before the real depositor did. With `2^62` possible child
+addresses, it is currently computationally infeasible to churn through the
+addresses fast enough to accomplish this.
+
+==== Commit + Reveal Scheme
+
+Now that we've establed that a particular derivation path like
+`m/44'/0/281'/0/1821912'` is a secret, then we can craft a commit+reveal scheme
+from it. First, the user deposits BTC to the public key at
+`m/44'/0/281'/0/1821912'`. Second, the user concatenates that secret path with their minting address:
+`m/44'/0/281'/0/1821912'+0x6EB149f6FBA44724cD2e7CAD13B18E5E8D2b709D`. This is their final secret.
+
+They hash that secret and get something like `keep btc deposit:
+924deb62fcb9780220c04ecab100b97ee1dea25a30771552d483d1c7fc7ef04a` (using
+sha256) and submit that to the ethereum chain.
+
+Then, once that commitment is on the blockchain, they reveal their secret in
+plain text:
+`m/44'/0/281'/0/1821912'+0x6EB149f6FBA44724cD2e7CAD13B18E5E8D2b709D` to the
+ethereum chain. Once the keep network sees the derivation path + minting
+address, it checks to see if there was a deposit at the supplied address and if
+the sha256 of the supplied data is an existing commitment. This prevents miners
+from reading the mempool and submitting their own message like
+`m/44'/0/281'/0/1821912'+0xMYADDRESS` to claim that the tBTC should be theirs.
+If everything checks out, then we're good!
+
+==== Sweeping The Wallet
+
+When a user submits their reveal, the wallet should move the funds from the
+secret derivation path to the main wallet located at `m/44'/0/0/0/0`. After
+this step, the system is good to mint the equivalent amount of tBTC to the
+supplied ethereum address in the reveal.
```

Comment:
Why do we need to sweep them? Can't we track UTXO and use them during redemption? Or maybe we do not want to track UTXOs and we do it for simplification? If we sweep, someone has to pay for sweeping transactions. Where is this cost included so that the peg can be maintained?